client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/HintedReplayingDecoder.java

```diff
@@ -0,0 +1,139 @@
+package org.infinispan.client.hotrod.impl.transport.netty;
+
+import java.util.Collections;
+import java.util.List;
+
+import org.infinispan.client.hotrod.logging.Log;
+import org.infinispan.client.hotrod.logging.LogFactory;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.DecoderException;
+import io.netty.util.Signal;
+
+/**
+ * Copy-paste of {@link io.netty.handler.codec.ReplayingDecoder} which is hinted to not attempt decoding unless enough
+ * bytes are read.
+ *
+ * The decoder does not expect pass different message up the pipeline, this is a terminal read operation.
+ */
+public abstract class HintedReplayingDecoder<S> extends ByteToMessageDecoder {
```

Comment:
I've used the state change wherever possible/appropriate, e.g. when multiple values are sent, these are unmarshalled and put into collection, updating the checkpoint. With big values, you need to have them in the buffer for one moment anyway. The hinted decoder just optimizes out some unmarshalling attempts that are deemed to fail.

Reply:
Yeah I have to admit, this is odd why we have our own impl of `ReplayingDecoder` in this case. Why doesn't this just extend `ReplayingDecoder` from netty? It seems to have a lot of the same copied code.

The reason I no longer used `ReplayingDecoder` is solely due to performance. The netty team rewrote the HTTP decoder in this way and saw 177% throughput increase for decoding [1].

`ReplayingDecoder` is by far a lot easier to use that doing it by only extending `ByteToMessageDecoder`. This code is actually doing the former though. If we do change it to the latter we have to unfortunately check bounds at each point you read from the bytes. Fortunately it is easy to install an additional frame decoder that only allows 1 byte down at a time [2], which  finds a lot of issues in tests. Disregard the mentions of 2 :) - missed updating those before committing.

[1] https://github.com/netty/netty/commit/fda8808210c4f4f092a93da5929df43062c3aa6c#diff-bf7a179a319b29e0262d5badaf3770e0
[2] https://github.com/infinispan/infinispan/blob/master/server/hotrod/src/test/java/org/infinispan/server/hotrod/transport/SingleByteFrameDecoderChannelInitializer.java#L20

Reply:
>  why we have our own impl of ReplayingDecoder

Should I be more verbose than "to not attempt decoding unless enough bytes are read"? The issue was found in `ClientEventsOOMTest` where each event has 40M: this arrives chunked in 512kB buffers (and initially shorter ones). The buffer is then read in marshalling about 60 times, every time just finding out that it's too short. The hinted decoder finds out that we're trying to read 40M chunk and does not pass it to marshalling until enough data are in the buffer.

The bounds checks are executed once you know how many bytes should be read next, and that is either just next byte (with variable-length number encoding) or we are reading an array and do the bounds check only once.

I am not opposing avoiding the exception-throwing approach, but I don't know if I could rewrite it more efficiently.

Reply:
Makes sense, and yes I wanted the more detailed explanation (I just glanced at the code) :) I now better understand the issue this is trying to solve now and is an interesting approach. TBH I kinda wonder if the netty guys might want to include changes like this into their `ReplayingDecoder` even.

And tbh looking at how little decoding there is going on, the non exception approach could be better, but I doubt it would be that much, since there are so few possible checkpoints. It would however complicate the code substantially. So I would say to keep going with this for now.